ambry-tools/src/main/java/com.github.ambry.tools/perf/rest/PerfRSC.java

```diff
@@ -0,0 +1,139 @@
+ * specified size.
+ */
+class PerfRSC implements ReadableStreamChannel {
+  private final ByteBuffer bytes;
+  private final long streamSize;
+  private final AtomicBoolean readIntoInvoked = new AtomicBoolean(false);
+  private final AtomicBoolean callbackInvoked = new AtomicBoolean(false);
+  private final AtomicBoolean isOpen = new AtomicBoolean(true);
+
+  private volatile long streamed = 0;
+  private volatile Exception channelWriteException = null;
+
+  private volatile AsyncWritableChannel writeChannel = null;
+  private volatile FutureResult<Long> futureResult = new FutureResult<Long>();
+  private volatile Callback<Long> readIntoCallback = null;
+
+  /**
+   * Create an instance of PerfRSC.
+   * @param bytes the bytes that form a chunk that will be repeatedly returned until {@code streamSize} is reached.
+   * @param streamSize the desired total size of the stream.
+   */
+  public PerfRSC(byte[] bytes, long streamSize) {
+    if (streamSize < 0 || (streamSize > 0 && bytes.length < 1)) {
+      throw new IllegalArgumentException("Invalid argument(s)");
+    }
+    this.streamSize = streamSize;
+    if (bytes != null) {
+      this.bytes = ByteBuffer.wrap(bytes);
+    } else {
+      this.bytes = null;
+    }
+  }
+
+  @Override
+  public long getSize() {
+    return streamSize;
+  }
+
+  @Override
+  public Future<Long> readInto(AsyncWritableChannel asyncWritableChannel, Callback<Long> callback) {
+    if (readIntoInvoked.compareAndSet(false, true)) {
+      writeChannel = asyncWritableChannel;
+      readIntoCallback = callback;
+      writeToChannel();
+    } else {
+      throw new IllegalStateException("ReadInto should not be called more than once");
+    }
+    return futureResult;
+  }
+
+  @Override
+  public void setDigestAlgorithm(String digestAlgorithm)
+      throws NoSuchAlgorithmException {
+    // no op
+  }
+
+  @Override
+  public byte[] getDigest() {
+    return null;
+  }
+
+  @Override
+  public boolean isOpen() {
+    return isOpen.get();
+  }
+
+  @Override
+  public void close() {
+    isOpen.set(false);
+  }
+
+  /**
+   * Writes continuously to the provided {@link AsyncWritableChannel} until {@link #streamSize} is reached.
+   */
+  private void writeToChannel() {
+    if (channelWriteException == null && streamed < streamSize) {
+      if (!isOpen()) {
+        invokeCallback(streamed, new ClosedChannelException());
+      } else {
+        int bytesLeft = streamSize - streamed > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) (streamSize - streamed);
+        bytes.rewind();
+        bytes.limit(Math.min(bytes.capacity(), bytesLeft));
+        writeChannel.write(bytes, new Callback<Long>() {
+          @Override
+          public void onCompletion(Long result, Exception callbackException) {
+            streamed += result;
+            channelWriteException = callbackException;
+            writeToChannel();
```

Comment:
Two reasons I think
1. Don't want to occupy the thread that did the `readInto()` call (which is the function that calls this one). The function is supposed to be non blocking and looping a lot of times takes some time.
2. This acts as back pressure. I don't want to overload the `AsyncWritableChannel` with more writes than it can handle.